### PR TITLE
Enable patch releases by fetching some components from existing distribution.

### PR DESCRIPTION
**Pull request by dblock**
Signed-off-by: dblock <dblock@amazon.com>

### Description

- Adds a new component type that specifies "dist" to pull from an existing distribution instead of building from source.
- Adds "patches" into the input manifest that includes a list of existing versions this distribution patches.

This allows build to produce a mixed distribution with multiple versions of components, e.g. 1.1.0 + some components built from source = 1.1.1. 

In the provided example this build fails trying to build `anomaly-detection`, which expects OpenSearch core 1.1.1, but we're giving it 1.1.0. The builds for those patched components will need to pickup the released dependency from maven explicitly.
 
### Issues Resolved

Closes https://github.com/opensearch-project/opensearch-build/issues/373.
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
@saratvemulapalli @peternied Bump, this is ready for review please.

File: manifests/1.1.1/opensearch-1.1.1.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+---
+schema-version: "1.0"
+build:
+  name: OpenSearch
+  version: 1.1.1
+  patches: 
+    - 1.1.0
+components:
+  - name: OpenSearch
+    dist: https://ci.opensearch.org/ci/dbc/builds/1.1.0/405
```
Comment (peternied) [inline manifests/1.1.1/opensearch-1.1.1.yml:10]: The path will need to be inferred/discovered which will have been broken with the recent url restructuring that only impacted the 1.2.0+ builds.  Should we add parameters to help the build system construct the correct path to the manifest?

**Review by peternied — COMMENTED** [review]
General feedback, deeper inspection later

File: manifests/1.1.1/opensearch-1.1.1.yml (hunk 1 above)
Comment (dblock) [reply manifests/1.1.1/opensearch-1.1.1.yml:10]: Either in a separate PR or maybe we should just copy the manifest with bad URLs to a corrected copy in S3 and use that instead.

File: src/build_workflow/build_target.py [hunk 2]
```diff
@@ -41,7 +43,22 @@ def __init__(
     def opensearch_version(self):
         return self.version + "-SNAPSHOT" if self.snapshot else self.version
```
Comment (peternied) [inline src/build_workflow/build_target.py:44]: Nearly all properties in this class have an `if` statement, code smells like we should create a subclass

File: src/build_workflow/builder_from_dist.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            self.ref = manifest.ref
+            self.sha = manifest.commit_id
+
+    def checkout(self, work_dir):
+        self.__download_build_manifest()
+
+    def build(self, build_recorder):
+        pass
+
+    def export_artifacts(self, build_recorder):
```
Comment (peternied) [inline src/build_workflow/builder_from_dist.py:28]: `export` doesn't seem to align with the action what about `import` or `download` + "_artifacts"?

File: manifests/1.1.1/opensearch-1.1.1.yml (hunk 1 above)
Comment (peternied) [reply manifests/1.1.1/opensearch-1.1.1.yml:10]: The manifest is oriented around the components of with a platform / architecture, do we need another kind of manifest that is oriented around a releases artifacts.  Note, we can move this out to another PR

Comment (dblock) [reply manifests/1.1.1/opensearch-1.1.1.yml:10]: I don't see this as necessary. This is a bundle (assembly/distribution) manifest that can now include references to another bundle, and picks up components with matching platform/architecture.

File: src/build_workflow/build_target.py (hunk 2 above)
Comment (dblock) [reply src/build_workflow/build_target.py:44]: https://github.com/opensearch-project/opensearch-build/issues/856

File: src/build_workflow/builder_from_dist.py (hunk 3 above)
Comment (dblock) [reply src/build_workflow/builder_from_dist.py:28]: I thought about this. The "from source" code exports those artifacts, while the "from distribution" code downoads and exports those artifacts. They should really be spit up into `download_artifacts` and `export_artifacts`. I have another PR with more changes along those lines if that's OK?

**Review by peternied — APPROVED** [review]

**Review by saratvemulapalli — APPROVED** [review]
Lots of changes. 
Did a pass, overall it looks good to me.

**Pull request merged**